lsp/nls/src/linearization/building.rs

```diff
@@ -120,40 +139,37 @@ impl<'b> Building<'b> {
 
     pub(super) fn resolve_reference<'a>(
         &'a self,
-        item: &'a LinearizationItem<UnifType>,
-    ) -> Option<&'a LinearizationItem<UnifType>> {
-        // if item is a usage, resolve the usage first
-        match item.kind {
-            TermKind::Usage(UsageState::Resolved(pointed)) => self.linearization.get(pointed.index),
-            _ => Some(item),
-        }
-        // load referenced value, either from record field or declaration
-        .and_then(|item_pointer| {
-            match &item_pointer.kind {
-                // if declaration is a record field, resolve its value
-                TermKind::RecordField { value, .. } => {
-                    debug!("parent referenced a record field {:?}", value);
-                    value
-                        // retrieve record
-                        .as_option()
-                        .and_then(|value_index| self.linearization.get(value_index.index))
-                }
-                // if declaration is a let binding resolve its value
-                TermKind::Declaration(_, _, ValueState::Known(value)) => {
-                    self.linearization.get(value.index)
-                }
+        current_file: FileId,
+        item: &'a TermKind,
+    ) -> Option<&'a TermKind> {
+        match item {
+            // if declaration is a record field, resolve its value
+            TermKind::RecordField { value, .. } => {
+                debug!("parent referenced a record field {:?}", value);
+                value
+                    // retrieve record
+                    .as_option()
+                    .and_then(|value_index| self.get_item_kind(current_file, value_index))
+            }
+            // if declaration is a let binding resolve its value
+            TermKind::Declaration(_, _, ValueState::Known(value)) => {
+                self.get_item_kind(current_file, *value)
+            }
 
-                // if something else was referenced, stop.
-                _ => Some(item_pointer),
+            TermKind::Usage(UsageState::Resolved(pointed)) => {
+                let kind = self.get_item_kind(current_file, *pointed)?;
+                self.resolve_reference(current_file, kind)
             }
-        })
+            // if something else was referenced, stop.
+            _ => Some(item),
+        }
     }
 
     pub(super) fn resolve_record_references(
         &mut self,
         file: FileId,
         mut defers: Vec<(ItemId, ItemId, Ident)>,
-    ) {
+    ) -> Vec<(ItemId, ItemId, Ident)> {
```

Comment:
What is this new return value doing?

Reply:
It returns a list of items that could not be resolved.

Reply:
I see. Then it might be valuable to document this function, and in particular what the return value is.